tests/integration/test.replication_events.js

```diff
@@ -0,0 +1,174 @@
+        // We should receive a paused event when replication
+        // starts because there is nothing to replicate
+        if (counter === 1) {
+          db.bulkDocs([{_id: 'a'}, {_id: 'b'}]);
+        } else if (counter === 3) {
+          db.bulkDocs([{_id: 'c'}, {_id: 'd'}]);
+        } else if (counter === 5) {
+          repl.cancel();
+        } else {
+          done('paused fired incorrectly');
+        }
+      });
+    });
+
+
+    it('#3852 Test basic starting with docs', function (done) {
+
+      var db = new PouchDB(dbs.name);
+
+      db.bulkDocs([{_id: 'a'}, {_id: 'b'}]).then(function() {
+
+        var repl = db.replicate.to(dbs.remote, {retry: true, live: true});
+
+        var counter = 0;
+
+        repl.on('complete', function() { done(); });
+
+        repl.on('active', function(evt) {
+          counter++;
+          if (!(counter === 1 || counter === 3 || counter === 5)) {
+            done('active fired incorrectly:' + counter);
+          }
+        });
+
+        repl.on('paused', function(evt) {
+          counter++;
+          // We should receive a paused event when replication
+          // starts because there is nothing to replicate
+          if (counter === 2) {
+            db.bulkDocs([{_id: 'c'}, {_id: 'd'}]);
+          } else if (counter === 4) {
+            db.bulkDocs([{_id: 'e'}, {_id: 'f'}]);
+          } else if (counter === 6) {
+            repl.cancel();
+          } else {
+            done('paused fired incorrectly');
+          }
+        });
+      });
+    });
+
+    it('#3852 Test errors', function (done) {
+
+      if (!(/http/.test(dbs.remote) && !/http/.test(dbs.name))) {
+        // Only run test when remote is http and local is local
+        return done();
+      }
+
+      var db = new PouchDB(dbs.name);
+      var rejectAjax = true;
+      var ajax = PouchDB.utils.ajax;
+
+      PouchDB.utils.ajax = function (opts, cb) {
+        if (rejectAjax) {
+          cb(new Error('flunking you'));
+        } else {
+          ajax.apply(this, arguments);
+        }
+      };
+
+      db.bulkDocs([{_id: 'a'}, {_id: 'b'}]).then(function() {
+
+        var repl = db.replicate.to(dbs.remote, {
+          retry: true,
+          live: true,
+          back_off_function: function () { return 0; }
+        });
+
+        var counter = 0;
+
+        repl.on('complete', function() {
+          PouchDB.utils.ajax = ajax;
+          done();
+        });
+
+        repl.on('active', function(evt) {
+          counter++;
+          if (counter === 2) {
+            // All good, wait for pause
+          } else if (counter === 4) {
+            // Lets start failing while active
+            rejectAjax = true;
+            db.bulkDocs([{_id: 'e'}, {_id: 'f'}]);
+          } else if (counter === 6) {
+            // All good, wait for pause
+          } else {
+            done('active fired incorrectly');
+          }
+        });
+
+        repl.on('paused', function(err) {
+          counter++;
```

Comment:
Its possible, I try to reserve sinon for actually mocking implemetation in unit tests, I find a lot of times when people put sinon in integration tests they hide fairly large errors, but will see how sinon is working in the other PR for the replication options